pkg/visitors/insert.go

```diff
@@ -121,7 +121,21 @@ func (v *Insert) Visit(f uefi.Firmware) error {
 				case InsertDXE:
 					fallthrough
 				case InsertEnd:
-					f.Files = append(f.Files, v.NewFile)
+					non_empty_pad := -1
+					for i := 0; i < len(f.Files); i++ {
+						fileBuf := f.Files[i].Buf()
+						fileLen := uint64(len(fileBuf))
+						if f.Files[i].Header.GUID == *uefi.FFGUID && fileLen > 24 {
+							// found non-empty PAD to check for free space.
+							non_empty_pad = i
+						}
+					}
+					if non_empty_pad >= 0 {
+						i := non_empty_pad
+						f.Files = append(f.Files[:i], append([]*uefi.File{v.NewFile}, f.Files[i:]...)...)
```

Comment:
Why are you inserting more files whenever you detect a non empty padfile? There are many bioses that have non empty pad files for alignment reasons.

Reply:
When I run out of space to insert EFI drivers in the the DXE FV for example, then they are inserted into the last FV and claims space from the non-empty pad file to insert them.  In my use model I do:
     if $UTK -xzPath ${XZ} ${SRC} insert_dxe $file save $DST
when that fails do running out of space in that FV then I do:
     if $UTK -xzPath ${XZ} ${SRC} insert_end PeiCore $file save $DST
I iterate through all of the EFI drivers.   After a successful insert then I move $DST to $SRC. I insert a EFI driver one at time so that when the DXE FV gets full then it tries the
last firmware volume identified via PeiCore.  However, there is no reason why multiple EFI drivers couldn't be installed at the same time.  Note that as EFI drivers are added the pad file is shrunk by the size of the EFI driver.  So then the modified BIOS shows the new EFI driver as a DXE driver in UEFITool and the the non-empy pad file smaller.